crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,770 @@
+    }
+}
+
+#[pymethods]
+impl CommutationLibrary {
+    #[new]
+    fn new(py_any: Option<Bound<PyAny>>) -> Self {
+        match py_any {
+            Some(pyob) => CommutationLibrary {
+                library: pyob
+                    .extract::<Option<HashMap<(String, String), CommutationLibraryEntry>>>()
+                    .unwrap(),
+            },
+            None => CommutationLibrary {
+                library: Some(HashMap::new()),
+            },
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum CommutationLibraryEntry {
+    Commutes(bool),
+    QubitMapping(HashMap<SmallVec<[Option<Qubit>; 2]>, bool>),
+}
+
+impl<'py> FromPyObject<'py> for CommutationLibraryEntry {
+    fn extract_bound(b: &Bound<'py, PyAny>) -> Result<Self, PyErr> {
+        if let Ok(b) = b.extract::<bool>() {
+            return Ok(CommutationLibraryEntry::Commutes(b));
+        }
+        let dict = b.downcast::<PyDict>()?;
+        let mut ret = hashbrown::HashMap::with_capacity(dict.len());
+        for (k, v) in dict {
+            let raw_key: SmallVec<[Option<u32>; 2]> = k.extract()?;
+            let v: bool = v.extract()?;
+            let key = raw_key.into_iter().map(|key| key.map(Qubit)).collect();
+            ret.insert(key, v);
+        }
+        Ok(CommutationLibraryEntry::QubitMapping(ret))
+    }
+}
+
+impl ToPyObject for CommutationLibraryEntry {
+    fn to_object(&self, py: Python) -> PyObject {
+        match self {
+            CommutationLibraryEntry::Commutes(b) => b.into_py(py),
+            CommutationLibraryEntry::QubitMapping(qm) => {
+                let out_dict = PyDict::new_bound(py);
+
+                qm.iter().for_each(|(k, v)| {
+                    out_dict
+                        .set_item(
+                            PyTuple::new_bound(py, k.iter().map(|q| q.map(|t| t.0))),
+                            PyBool::new_bound(py, *v),
+                        )
+                        .ok()
+                        .unwrap()
+                });
+                out_dict.unbind().into_any()
+            }
+        }
+    }
+}
+
+type CacheKey = (
+    SmallVec<[Option<Qubit>; 2]>,
+    (SmallVec<[ParameterKey; 3]>, SmallVec<[ParameterKey; 3]>),
+);
+
+type CommutationCacheEntry = HashMap<CacheKey, bool>;
+
+fn commutation_entry_to_pydict(py: Python, entry: &CommutationCacheEntry) -> Py<PyDict> {
+    let out_dict = PyDict::new_bound(py);
+    for (k, v) in entry.iter() {
+        let qubits = PyTuple::new_bound(py, k.0.iter().map(|q| q.map(|t| t.0)));
+        let params0 = PyTuple::new_bound(py, k.1 .0.iter().map(|pk| pk.0));
+        let params1 = PyTuple::new_bound(py, k.1 .1.iter().map(|pk| pk.0));
+        out_dict
+            .set_item(
+                PyTuple::new_bound(py, [qubits, PyTuple::new_bound(py, [params0, params1])]),
+                PyBool::new_bound(py, *v),
+            )
+            .expect("Failed to construct commutation cache for serialization");
+    }
+    out_dict.unbind()
+}
```

Comment:
Since this appears to be fallible. You should use a PyResult as the return value here to avoid panicking.

```suggestion
fn commutation_entry_to_pydict(py: Python, entry: &CommutationCacheEntry) -> PyResult<Py<PyDict>> {
    let out_dict = PyDict::new_bound(py);
    for (k, v) in entry.iter() {
        let qubits = PyTuple::new_bound(py, k.0.iter().map(|q| q.map(|t| t.0)));
        let params0 = PyTuple::new_bound(py, k.1 .0.iter().map(|pk| pk.0));
        let params1 = PyTuple::new_bound(py, k.1 .1.iter().map(|pk| pk.0));
        out_dict
            .set_item(
                PyTuple::new_bound(py, [qubits, PyTuple::new_bound(py, [params0, params1])]),
                PyBool::new_bound(py, *v),
            )?;
    }
    Ok(out_dict.unbind())
}
```

Reply:
Done in 12d0a05